src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -91,7 +81,7 @@ public override void AdvanceTo(SequencePosition consumed, SequencePosition exami
 
         private void AdvanceTo(BufferSegment consumedSegment, int consumedIndex, BufferSegment examinedSegment, int examinedIndex)
         {
-            if (consumedSegment == null || examinedSegment == null)
+            if (consumedSegment == null || examinedSegment == null || !_streamReadTask.IsCompleted)
```

Comment:
How is the reader getting a non-null consumedSegment/examinedSegment for an unfinished/canceled read anyway? Can we at least do some validation to verify the caller isn't trying to actual advance the cursor in this case?

Reply:
Because we return the existing buffer for cancelled reads. You could have buffered data on a previous read then cancelled the next read.

Reply:
Exactly, and you're just ignoring that without advancing the read head and returning. You should add a test for this.

Reply:
For what exactly?

Reply:
Calling AdvanceTo with a real cursor during an ongoing call to InnerStream.ReadAsyc. It's super broken right now. If you don't want to provid solid support it, at least throw in that case.

Reply:
Just so I’m clear, the scenario is I have data buffered, and I cancel the read and call Advance with a real cursor? Yes that seems like something that should be handled. I just need to make sure to we don’t clear the memory in the pending read

Reply:
Yep. That's the scenario I was thinking of.

Reply:
Sounds good